tests/sharing_test.py

```diff
@@ -0,0 +1,166 @@
+        ad_token, keys = self.setup_sharing_and_encrypt()
+        contains_base_64_special_chars = "+" in ad_token or "/" in ad_token or "=" in ad_token
+        self.assertFalse(contains_base_64_special_chars)
+
+    def test_uid2_client_produces_uid2_token(self):
+        ad_token, keys = self.setup_sharing_and_encrypt()
+        self.assertEqual("A", ad_token[0])
+
+    def test_euid_client_produces_euid_token(self):
+        ad_token, keys = self.setup_sharing_and_encrypt(IdentityScope.EUID)
+        self.assertEqual("E", ad_token[0])
+
+    def _get_token_identity_type(self, uid2, keys):
+        token = encrypt_key(uid2, IdentityScope.UID2, keys)
+
+        first_char = token[0]
+        if ('A' == first_char or 'E' == first_char):
+            return IdentityType.Email
+        if ('F' == first_char or 'B' == first_char):
+            return IdentityType.Phone
+
+        return None
+
+    def test_raw_uid_produces_correct_identity_type_in_token(self):
+        ad_token, keys = self.setup_sharing_and_encrypt()
+
+        self.assertEqual(IdentityType.Email, self._get_token_identity_type("Q4bGug8t1xjsutKLCNjnb5fTlXSvIQukmahYDJeLBtk=", keys))
+        self.assertEqual(IdentityType.Phone, self._get_token_identity_type("BEOGxroPLdcY7LrSiwjY52+X05V0ryELpJmoWAyXiwbZ", keys))
+        self.assertEqual(IdentityType.Email, self._get_token_identity_type("oKg0ZY9ieD/CGMEjAA0kcq+8aUbLMBG0MgCT3kWUnJs=", keys))
+        self.assertEqual(IdentityType.Email, self._get_token_identity_type("AKCoNGWPYng/whjBIwANJHKvvGlGyzARtDIAk95FlJyb", keys))
+        self.assertEqual(IdentityType.Email, self._get_token_identity_type("EKCoNGWPYng/whjBIwANJHKvvGlGyzARtDIAk95FlJyb", keys))
+
+    def test_multiple_keys_per_keyset(self):
+        master_key2 = EncryptionKey(264, -1, _now - dt.timedelta(days=-2), _now - dt.timedelta(days=-1), _now - dt.timedelta(hours=-1),
+                            _master_secret, keyset_id=9999)
+        site_key2 = EncryptionKey(_site_key_id, _site_id, _now - dt.timedelta(days=-2), _now - dt.timedelta(days=-1), _now - dt.timedelta(hours=-1),
+                          _site_secret, keyset_id=99999)
+        client = Uid2Client("endpoint", "authkey", _client_secret)
+        json_body = self._key_set_to_json_for_sharing([_master_key, master_key2, _site_key, site_key2])
+        keys = client.refresh_json(json_body)
+
+        ad_token = encrypt_key(_example_uid, IdentityScope.UID2, keys)
+
+        result = decrypt_token(ad_token, keys)
+
+        self.assertEqual(_example_uid, result.uid2)
+
+    def test_cannot_encrypt_if_no_key_from_default_keyset(self):
+        client = Uid2Client("endpoint", "authkey", _client_secret)
+        json_body = self._key_set_to_json_for_sharing([_master_key])
+        keys = client.refresh_json(json_body)
+
+        self.assertRaises(EncryptionError, encrypt_key, _example_uid, IdentityScope.UID2, keys)
+
+    def test_expiry_in_token_matches_expiry_in_reponse(self):
```

Comment:
Are we missing CannotEncryptIfTheresNoDefaultKeysetHeader()? That one's relevant because there's been talk of splitting SHARER into SHARER_SENDER vs SHARER_RECEIVER